### PR TITLE
[SAMZA-2557] Adding support for nested rows access via dot path.

### PR DESCRIPTION
**Pull request by b-slim**
This is an initial draft on how to support the nested row access in Samza SQL.
There are multiple  #interconnected items.

1. Added a the actual definition of a ROW Calcite Data Type [**EASY FINAL**].

2. Added a Row Converter form Samza Type System to Calcite Type System [**Okay for now but will need more work for types like timestamps**].
3. Added a Collector for projects and filters that are pushed to Remote Table Scan **[Complex and Needs Discussions**]. 

 - Why we need this ? Adding a nested row struct forces the addition of project and in general nothing stops Calcite logical planner to add such an identity project thus this is needed anyway.

- How this done ? As of now I chose to minimize the amount of rewrite or refactor and added a queue to collect the call stack between Remote table Scan and Join node. Then When doing the join The Project and Filter will happen post Join Lookup. We need to handle the case where filter does not match and null pad the result or return null as by current convention. To be honest I am still debating adding the Filter push down seems like there is no real gain since we have done the lookup already.

4. Added Ref Implementation for the getNestedField legacy udf (This udf has to be removed because it assumes Everything is a SamzaRelRecord) [**DONE**].
5. Need more code cleaning where type is mixed up between String Java and Avro Utf8 Java as a Key in the map [**Follow up**].
6. Need more work on the union Type System case we have more than 2 Types [**Follow up**].

---
This change is [Reviewable](https://reviewable.io/reviews/apache/samza/1386)

### CODE REVIEW

File: samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java [hunk 1]
```diff
@@ -141,20 +142,28 @@ public RelRoot plan(String query) {
           .sqlToRelConverterConfig(SqlToRelConverter.Config.DEFAULT)
           .traitDefs(traitDefs)
           .context(Contexts.EMPTY_CONTEXT)
-          .costFactory(null)
+          .programs(
+              Programs.hep(ImmutableList.of(FilterJoinRule.FILTER_ON_JOIN), true, DefaultRelMetadataProvider.INSTANCE))
```
Comment (atoomula) [inline samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java:146]: Could we remove FILTER_ON_JOIN optimization ? It doesn't work well with remote joins. We should instead use the optimization for remote joins in the other PR that I sent.

Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java:146]: @atoomula this rule is not really the main part of this work and I don't think we need to fix/choose now which rule should be used to handle filters or should or should not be pushed to table scan this is beyond the scope of this PR. 
The main goal is to explore how we can handle the **operator stack between the join and remote table scan**. 
I added the rule to ensure that if a filter is there things works as expected and it seems it is working. But I would love to know what are the **blind spot(s)** I am missing about what work and does not work well with the remote table scan if you think this will help drive this issue. Thanks

File: samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+  @Override
+  public MessageStream<SamzaSqlRelMessage> filter(FilterFunction<? super SamzaSqlRelMessage> filterFn) {
+    _mapFnCallQueue.offer(new FilterMapAdapter(filterFn));
+    return this;
+  }
+
+   Function<SamzaSqlRelMessage, SamzaSqlRelMessage> getFunction(Context context) {
+    Function<SamzaSqlRelMessage, SamzaSqlRelMessage> tailFn = null;
+    while (!_mapFnCallQueue.isEmpty()) {
```
Comment (srinipunuru) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:75]: Can you add more comments on what we are doing here?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+      if (tailFn == null) {
+        tailFn = current;
+      } else {
+        tailFn = current.compose(tailFn);
+      }
+    }
+    return tailFn == null ? Function.identity() : tailFn;
+  }
+
+  private static class FilterMapAdapter implements MapFunction<SamzaSqlRelMessage, SamzaSqlRelMessage> {
```
Comment (srinipunuru) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:94]: Can you add comments here as well on why this adapter is required and what it does?

File: samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java (hunk 2 above)
Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:93]: @srinipunuru added comments at the top of the class let me know if this still unclear.

File: samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java (hunk 3 above)
Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:120]: Added comments let me know if it is still unclear.

File: samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+import org.apache.samza.table.Table;
+
+
+/**
+ * Collector of Map and Filter Samza Functions to collect call stack on the top of Remote table.
+ * This Collector will be used by Join operator and trigger it when applying the join function post lookup.
+ *
+ * Note that this is needed because the Remote Table can not expose a proper {@code MessageStream}.
+ * It is a work around to minimize the amount of code changes of the current Query Translator {@link org.apache.samza.sql.translator.QueryTranslator},
+ * But in an ideal world, we should use Calcite planner in conventional way we can combine function when via translation of RelNodes.
```
Comment (atoomula) [inline samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:54]: I still do not get it. Why can't we use query optimization for remote tables and if we see a filter/projection between table (we can detect table vs stream in the optimizer rule) and join in the Calcite plan, push them up ?

Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:54]: The conventional way of using Calcite is pushing operator toward table scan as much as possible. 
This PR does that by leveraging Calcite as is with minimal code changes or copy and past of some internal code. 
By pushing some stuff up case remote table and some stuff down case remote table shows a clear disconnect and not clear design that a someone else beside the author of the work will be able to get it without spending hours stepping into the debugger code. 
The other question how this will be better than current approach of composing map/filters and fuse it to join operator, don't you think it is more optimal to fuse all the lookup/project/filter within join as one operator ? 
Adding to that This way also, will enable a case where the join condition will be more than just __key__ = c but will be able to add more conjunctions in the near future.  
In my opinion this is the most clean way to work around the limitation of the remote table join operator with no major surgery and allowing pushing filters/projects and in the future handle richer join conditions. 
Again the proper fix will be to adopt Calcite framework Convention pattern where an operator can be pushed inside another operator when going from logical to physical for instance in this case the Join will transformed to a new join node that knows how to translate the project and filters within it self. That is kind of what is happening now but without making this work a major surgery as someone has suggested.

Comment (atoomula) [reply samza-sql/src/main/java/org/apache/samza/sql/translator/MessageStreamCollector.java:54]: Ahh.. now I get it. If we have the following condition "(p.__key__ + 1) = pv.profileId", we cannot really solve it with calcite rule change other than changing our translator code (one of which is what you did).

File: samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+import org.apache.calcite.sql.validate.SqlUserDefinedFunction;
+
+import static org.apache.calcite.schema.impl.ReflectiveFunctionBase.builder;
+
+
+/**
+ * Operator to extract nested Rows or Fields form a struct row type using a dotted path.
+ * The goal of this operator is two-fold.
+ * First it is a temporary fix for https://issues.apache.org/jira/browse/CALCITE-4065 to extract a row from a row.
+ * Second it will enable smooth backward compatible migration from existing udf that relies on legacy row format.
```
Comment (atoomula) [inline samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java:62]: Is this truly backward compatible with with the existing GetNestedField udf ? Does this support all the types that are tested in GetSqlFieldUdf ? Esp map and array. https://github.com/apache/samza/blob/dcd4b558a2c702f5b5a320fdb9d0c3fcadabd09b/samza-sql/src/test/java/org/apache/samza/sql/fn/TestGetSqlFieldUdf.java

File: samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java [hunk 6]
```diff
@@ -194,10 +194,12 @@ public static Object convertToAvroObject(Object relObj, Schema schema) {
             .collect(Collectors.toList());
         return avroList;
       case MAP:
-        return ((Map<String, ?>) relObj).entrySet()
-            .stream()
-            .collect(Collectors.toMap(Map.Entry::getKey,
-              e -> convertToAvroObject(e.getValue(), getNonNullUnionSchema(schema).getValueType())));
+        // If you ask why not using String and that is because some strings are Wrapped into org.apache.avro.util.Utf8
+        // TODO looking at the Utf8 code base it is not immutable, having it as a key is calling for trouble!
+        final Map<Object, Object> outputMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java:199]: @atoomula The code as of today does that, in fact the reason I run into it is because I changed to String but tested failed. To avoid making this PR a fix everything pr I added the comment and base fix for the null case needed by my work. But I agree we need to move out of Avro String especially that is is mutable.

File: samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java (hunk 5 above)
Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java:62]: It should be I run all the tests and it passed, if you have any test or use case in mind please bring it up. But this should work with all the type and will enforce type checking as oppose to the old udf that is type less in bunch corner cases. In nutshell some failure will happen if type does not match.

File: samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java [hunk 7]
```diff
@@ -110,46 +114,50 @@ private void registerSourceSchemas(SchemaPlus rootSchema) {
...
-          .defaultSchema(rootSchema)
-          .operatorTable(new ChainedSqlOperatorTable(sqlOperatorTables))
-          .sqlToRelConverterConfig(SqlToRelConverter.Config.DEFAULT)
-          .traitDefs(traitDefs)
-          .context(Contexts.EMPTY_CONTEXT)
-          .costFactory(null)
-          .build();
-      Planner planner = Frameworks.getPlanner(frameworkConfig);
-
+    SchemaPlus rootSchema = CalciteSchema.createRootSchema(true, false).plus();
```
[the remark this answers is not in the document]
Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/planner/QueryPlanner.java:117]: We do not need caching, it is a historic optimization flag that most of cases not really needed because schema changes all the time. In fact that is the default in Calcite.

File: samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java (hunk 5 above)
Comment (atoomula) [reply samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java:62]: In the above test TestGetSqlFieldUdf, can you take a look to see if it handles tests starting from testMapAtLastField to testArrayAtAllIntermediateFields ? I don't think we have any tests in TestSamzaSqlEndToEnd that test such complex types at intermediate fields.

Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java:62]: Added more tests for this to mimic the unit tests. As you can see this will not include on how legacy handles map. Thus will need for sure to re-write such statement when going with new releases.

Comment (b-slim) [reply samza-sql/src/main/java/org/apache/samza/sql/udf/GetNestedField.java:62]: see https://github.com/apache/samza/pull/1386/commits/3e95902b615bd90ec686902fadbab9f98c03a75c
